File: sdk/go/files.go
package axernsdk

import (
	"context"

	filev1 "github.com/cofy-x/axern/sdk/go/gen/axern/common/file/v1"
)

// SandboxFileKind describes the kind of a sandbox filesystem entry.
type SandboxFileKind string

const (
	SandboxFileKindFile        SandboxFileKind = "file"
	SandboxFileKindDirectory   SandboxFileKind = "directory"
	SandboxFileKindSymlink     SandboxFileKind = "symlink"
	SandboxFileKindOther       SandboxFileKind = "other"
	SandboxFileKindUnspecified SandboxFileKind = "unspecified"
)

// SandboxFileInfo contains sandbox filesystem metadata.
type SandboxFileInfo struct {
	Path    string
	Kind    SandboxFileKind
	Size    int64
	Mode    uint32
	MtimeNS int64
}

// WriteFileOptions configures sandbox file writes.
type WriteFileOptions struct {
	CreateParents bool
}

// MkdirOptions configures sandbox directory creation.
type MkdirOptions struct {
	Parents bool
}

// RemoveOptions configures sandbox file or directory removal.
type RemoveOptions struct {
	Recursive bool
	Force     bool
}

// CopyOptions configures sandbox-side copy operations.
type CopyOptions struct {
	Recursive bool
	Overwrite bool
}

// MoveOptions configures sandbox-side move operations.
type MoveOptions struct {
	Overwrite bool
}

// ChmodOptions configures sandbox-side chmod operations.
type ChmodOptions struct {
	Recursive bool
}

// TouchOptions configures sandbox-side touch operations.
type TouchOptions struct {
	NoCreate bool
	MtimeNS  int64
}

// Stat returns metadata for a sandbox path.
func (s *Sandbox) Stat(ctx context.Context, path string) (SandboxFileInfo, error) {
	node, err := s.nodeClient()
	if err != nil {
		return SandboxFileInfo{}, err
	}
	return node.Stat(ctx, path)
}

// ListDir returns direct entries for a sandbox directory.
func (s *Sandbox) ListDir(ctx context.Context, path string) ([]SandboxFileInfo, error) {
	node, err := s.nodeClient()
	if err != nil {
		return nil, err
	}
	return node.ListDir(ctx, path)
}

// Exists reports whether a sandbox path exists.
func (s *Sandbox) Exists(ctx context.Context, path string) (bool, error) {
	node, err := s.nodeClient()
	if err != nil {
		return false, err
	}
	return node.Exists(ctx, path)
}

// ReadFile reads a sandbox file as bytes.
func (s *Sandbox) ReadFile(ctx context.Context, path string) ([]byte, error) {
	node, err := s.nodeClient()
	if err != nil {
		return nil, err
	}
	return node.ReadFile(ctx, path)
}

// WriteFile writes bytes to a sandbox file.
func (s *Sandbox) WriteFile(ctx context.Context, path string, data []byte, options WriteFileOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.WriteFile(ctx, path, data, options)
}

// Mkdir creates a sandbox directory.
func (s *Sandbox) Mkdir(ctx context.Context, path string, options MkdirOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.Mkdir(ctx, path, options)
}

// Remove removes a sandbox file or directory.
func (s *Sandbox) Remove(ctx context.Context, path string, options RemoveOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.Remove(ctx, path, options)
}

// Copy copies a sandbox file or directory.
func (s *Sandbox) Copy(ctx context.Context, srcPath, dstPath string, options CopyOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.Copy(ctx, srcPath, dstPath, options)
}

// Move moves a sandbox file or directory.
func (s *Sandbox) Move(ctx context.Context, srcPath, dstPath string, options MoveOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.Move(ctx, srcPath, dstPath, options)
}

// Chmod changes sandbox file permissions.
func (s *Sandbox) Chmod(ctx context.Context, path string, mode uint32, options ChmodOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.Chmod(ctx, path, mode, options)
}

// Touch updates sandbox file timestamps, optionally creating the path.
func (s *Sandbox) Touch(ctx context.Context, path string, options TouchOptions) error {
	node, err := s.nodeClient()
	if err != nil {
		return err
	}
	return node.Touch(ctx, path, options)
}

func sandboxFileInfo(info *filev1.SandboxFileInfo) SandboxFileInfo {
	if info == nil {
		return SandboxFileInfo{Kind: SandboxFileKindUnspecified}
	}
	return SandboxFileInfo{
		Path:    info.GetPath(),
		Kind:    sandboxFileKind(info.GetKind()),
		Size:    info.GetSize(),
		Mode:    info.GetMode(),
		MtimeNS: info.GetMtimeNs(),
	}
}

func sandboxFileInfos(entries []*filev1.SandboxFileInfo) []SandboxFileInfo {
	if len(entries) == 0 {
		return nil
	}
	infos := make([]SandboxFileInfo, 0, len(entries))
	for _, entry := range entries {
		infos = append(infos, sandboxFileInfo(entry))
	}
	return infos
}

func sandboxFileKind(kind filev1.SandboxFileKind) SandboxFileKind {
	switch kind {
	case filev1.SandboxFileKind_SANDBOX_FILE_KIND_FILE:
		return SandboxFileKindFile
	case filev1.SandboxFileKind_SANDBOX_FILE_KIND_DIRECTORY:
		return SandboxFileKindDirectory
	case filev1.SandboxFileKind_SANDBOX_FILE_KIND_SYMLINK:
		return SandboxFileKindSymlink
	case filev1.SandboxFileKind_SANDBOX_FILE_KIND_OTHER:
		return SandboxFileKindOther
	default:
		return SandboxFileKindUnspecified
	}
}

File: sdk/typescript/src/node/files.ts
/**
 * @license
 * Copyright 2026 cofy-x
 * SPDX-License-Identifier: Apache-2.0
 */

import { mapRpcError } from "../errors/index.js";
import { unary } from "../generated/proto.js";
import type {
  ChmodOptions,
  CopyOptions,
  MkdirOptions,
  MoveOptions,
  RemoveOptions,
  SandboxFileInfo,
  TouchOptions,
  WriteFileOptions,
} from "../types.js";
import { nonEmptyPath } from "../validation.js";
import type { NodeClientContext } from "./context.js";

export async function stat(ctx: NodeClientContext, path: string): Promise<SandboxFileInfo> {
  const response = await fileUnary(ctx, "StatFile", { path: nonEmptyPath(path) });
  return fileInfo(response.info as Record<string, unknown> | undefined);
}

export async function listDir(ctx: NodeClientContext, path: string): Promise<SandboxFileInfo[]> {
  const response = await fileUnary(ctx, "ListDir", { path: nonEmptyPath(path) });
  return ((response.entries as Record<string, unknown>[] | undefined) ?? []).map(fileInfo);
}

export async function exists(ctx: NodeClientContext, path: string): Promise<boolean> {
  const response = await fileUnary(ctx, "Exists", { path: nonEmptyPath(path) });
  return response.exists === true;
}

export async function readFile(ctx: NodeClientContext, path: string): Promise<Buffer> {
  const response = await fileUnary(ctx, "ReadFile", { path: nonEmptyPath(path) });
  return Buffer.from((response.data as Buffer | Uint8Array | undefined) ?? []);
}

export async function writeFile(
  ctx: NodeClientContext,
  path: string,
  data: Buffer | Uint8Array | string,
  options: WriteFileOptions = {},
): Promise<void> {
  await fileUnary(ctx, "WriteFile", {
    path: nonEmptyPath(path),
    data: Buffer.isBuffer(data) ? data : Buffer.from(data),
    create_parents: options.createParents ?? false,
  });
}

export async function mkdir(ctx: NodeClientContext, path: string, options: MkdirOptions = {}): Promise<void> {
  await fileUnary(ctx, "Mkdir", { path: nonEmptyPath(path), parents: options.parents ?? false });
}

export async function remove(ctx: NodeClientContext, path: string, options: RemoveOptions = {}): Promise<void> {
  await fileUnary(ctx, "Remove", {
    path: nonEmptyPath(path),
    recursive: options.recursive ?? false,
    force: options.force ?? false,
  });
}

export async function copy(
  ctx: NodeClientContext,
  srcPath: string,
  dstPath: string,
  options: CopyOptions = {},
): Promise<void> {
  await fileUnary(ctx, "Copy", {
    src_path: nonEmptyPath(srcPath),
    dst_path: nonEmptyPath(dstPath),
    recursive: options.recursive ?? false,
    overwrite: options.overwrite ?? false,
  });
}

export async function move(
  ctx: NodeClientContext,
  srcPath: string,
  dstPath: string,
  options: MoveOptions = {},
): Promise<void> {
  await fileUnary(ctx, "Move", {
    src_path: nonEmptyPath(srcPath),
    dst_path: nonEmptyPath(dstPath),
    overwrite: options.overwrite ?? false,
  });
}

export async function chmod(ctx: NodeClientContext, path: string, mode: number, options: ChmodOptions = {}): Promise<void> {
  await fileUnary(ctx, "Chmod", {
    path: nonEmptyPath(path),
    mode,
    recursive: options.recursive ?? false,
  });
}

export async function touch(ctx: NodeClientContext, path: string, options: TouchOptions = {}): Promise<void> {
  await fileUnary(ctx, "Touch", {
    path: nonEmptyPath(path),
    create: options.create ?? true,
    mtime_ns: options.mtimeNs ?? 0,
  });
}

async function fileUnary(ctx: NodeClientContext, method: string, payload: Record<string, unknown>): Promise<Record<string, unknown>> {
  try {
    return await ctx.withAuthRetry(300, (client) =>
      unary(client, method, ctx.authRequest(payload)),
    );
  } catch (error) {
    throw mapRpcError(error, `sandbox ${method}`, ctx.allocationId);
  }
}

function fileInfo(input: Record<string, unknown> | undefined): SandboxFileInfo {
  return {
    path: String(input?.path ?? ""),
    kind: fileKind(input?.kind),
    size: Number(input?.size ?? 0),
    mode: Number(input?.mode ?? 0),
    mtimeNs: Number(input?.mtime_ns ?? 0),
  };
}

function fileKind(kind: unknown): SandboxFileInfo["kind"] {
  switch (kind) {
    case 1:
      return "file";
    case 2:
      return "directory";
    case 3:
      return "symlink";
    case 4:
      return "other";
    default:
      return "unspecified";
  }
}
